src/Func.cpp

```diff
@@ -197,23 +197,34 @@ void Func::define_extern(const std::string &function_name,
 
 /** Get the types of the buffers returned by an extern definition. */
 const Type &Func::output_type() const {
-    user_assert(defined())
-        << "Can't access output buffer of undefined Func.\n";
-    user_assert(func.output_types().size() == 1)
-        << "Can't call Func::output_type on Func \"" << name()
-        << "\" because it returns a Tuple.\n";
-    return func.output_types()[0];
+    const auto &types = defined() ? func.output_types() : func.required_types();
+    if (types.empty()) {
+        user_error << "Can't call Func::output_type on Func \"" << name()
+                   << "\" because it is undefined or has no type requirements.\n";
+    } else if (types.size() > 1) {
+        user_error << "Can't call Func::output_type on Func \"" << name()
+                   << "\" because it returns a Tuple.\n";
+    }
+    return types[0];
 }
 
 const std::vector<Type> &Func::output_types() const {
-    user_assert(defined())
-        << "Can't access output buffer of undefined Func.\n";
-    return func.output_types();
+    const auto &types = defined() ? func.output_types() : func.required_types();
+    user_assert(!types.empty())
+        << "Can't call Func::output_type on Func \"" << name()
+        << "\" because it is undefined or has no type requirements.\n";
+    return types;
 }
 
 /** Get the number of outputs this function has. */
 int Func::outputs() const {
-    return func.outputs();
+    const auto &types = defined() ? func.output_types() : func.required_types();
+    // For backwards compatibility, we want to return 0 if the Func
```

Comment:
Is there existing code that depends on this?

Reply:
Not that I know of, but Hyrum's Law exists

Reply:
(I mean, we can make it an error case and see what breaks...)

Reply:
I'd make it an error. Checking the number of outputs of an undefined untyped Func seems like something that should throw an error.

Reply:
ok. Should the same be true for `dimensions()`?

Reply:
I think so, yes.

Reply:
done and done